wip on multiPXD files support

Tolsi opened the pull request:
It should resolve #3

I have done successfully splitting large PXD files into many smaller ones, but the function from .dll cannot successfully convert them to WAV. When comparing, old PXDs in different files and these PXDs from one big one have differences and may require different function parameters or approaches when unpacking.

Tolsi committed:
wip on multipxd support

Tolsi committed:
Trying to call ADecompress

Tolsi committed:
There's no init and close now

Tolsi committed:
multipxd decompress was realized

Tolsi committed:
decompress multipxe files was realized

Tolsi commented:
ps why don't we include the .dll files in the repository? I think no one cares, but it will comfortable for users.

I did an unpacking of large files, but I wanted to finish a couple more things today / tomorrow. I will mark the PR as ready for review when ready

paator commented:
If someone wants to unpack .pxd, they probably have .dll file included in their eJay installation folder as well.
Considering the legal reasons, I would simply avoid including it.

Thank you for your hard work!

Tolsi committed:
Final changes

Tolsi commented:
It's ready now!

> Considering the legal reasons, I would simply avoid including it.

We extract the proprietary format to pirate audio samples, what are the ethical issues 🤣

Tolsi committed:
Update README.md

Tolsi committed:
Missed type name if empty in header is fixed

Tolsi committed:
A lot of bug fixes

now it support .scl files also

Tolsi commented:
Fixed a lot of issues and added .scl files support (for eJay 7, Hip Hop 6, and R&B packages). Maybe rename the app to eJay Decompresser? It's not just PXD converter now

Tolsi committed:
scl conversion path fix

paator commented on PXDConverter/MarshalService.cs:
```diff
@@ -1,26 +1,33 @@
 
 namespace PXDConverter
 {
+    interface PXDDecompressLibrary
```
I would move all new interfaces/classes/enums to separate files.
That would make the code more readable when it grows

Tolsi replied:
I will delete this interface because it is useless already

paator commented on PXDConverter/MarshalService.cs:
```diff
@@ -32,32 +39,58 @@ public static void InitializeDll()
         }
 
-        public static void ConvertWavToRawDataBuffer(string wavPath, string bufferTmpPath)
+        public void Decompress(string pxdPath, int leftOffset, int leftSize, int rightOffset, int rightSize, int sampleRate, string outputPath)
```
Are parameters: int leftOffset, int leftSize, int rightOffset, int rightSize, int sampleRate, string outputPath
actually needed? It seems that you don't use them in this method scope

Tolsi replied:
it was from the interface, will remove them

paator commented on PXDConverter/Program.cs:
```diff
@@ -28,132 +46,549 @@ private static void Main(string[] args)
-            Console.WriteLine("Searching for all .pxd files in specified location...");
+            bool isWave = ASCIIEncoding.ASCII.GetString(bytes[0..4]) == "RIFF";
+            // rest is useless if it's wav file, but anyway
+            int p = 4;
```
What's "p"? I assume it's "pointer", but I would simply give it a more appropriate name ;)
you use that name in several places.
what do you think?

Tolsi replied:
p like position, will rename this

paator commented on PXDConverter/Program.cs:
```diff
@@ -28,132 +46,549 @@ private static void Main(string[] args)
-                argPath = Directory.GetCurrentDirectory();
-            
-            string[] filePaths = Directory.GetFiles(argPath, "*.pxd", SearchOption.AllDirectories);
+        private static (int, R) processBytes<R>(byte[] arr, int p, int size, Func<byte[], R> convert)
```
TResult would be a more appropriate name for the generic parameter I think

paator commented on PXDConverter/Program.cs:
```diff
@@ -28,132 +46,549 @@ private static void Main(string[] args)
+            (p, byte id) = processBytes(descriptorBytes, p, 1, b => b[0]);
+            (p, byte group) = processBytes(descriptorBytes, p, 1, b => b[0]);
+            p += 2; // serarator
+            (p, short nameLenght) = processBytes(descriptorBytes, p, 2, b => BitConverter.ToInt16(b));
```
If I'm not wrong, every lambda like this one can be shortened to method group

Tolsi replied:
just `BitConverter.ToInt16` doesn't work here

paator reviewed (changes requested):
Great work!
Dropped a few comments, but these are small fixes that would make code more readable.
I'm really impressed by your hard work - it seems that I've only scratched the surface with my code.

Also the suggestion for changing repository name is great! :)

Tolsi committed:
code review fixes

Tolsi committed:
Useless interface was removed

and uselsess params

Tolsi commented:
To be honest, I have achieved my goals and do not plan to maintain this code anymore, feel free to improve the PR to the level necessary to merge it with the application. Thanks for the great starting point, I hope someone else will be able to please and find this code useful.

paator commented:
Thank you for your help!:)
